Detectors/TPC/reconstruction/include/TPCReconstruction/ClustererTask.h

```diff
@@ -8,112 +8,71 @@
-    Clusterer* getClusterer(ClustererType type) {
-      switch (type) {
-        case ClustererType::HW:   return mHwClusterer.get();
-        case ClustererType::Box:  return mBoxClusterer.get();
-      };
-    };
-
-    /// Returns pointer to Box Clusterer
-    /// \return  Pointer to Clusterer, nullptr if Clusterer was not enabled during Init()
-    BoxClusterer* getBoxClusterer()   const { return mBoxClusterer.get(); };
-    
-    /// Returns pointer to Hw Clusterer
-    /// \return  Pointer to Clusterer, nullptr if Clusterer was not enabled during Init()
-    HwClusterer* getHwClusterer()     const { return mHwClusterer.get(); };
-
-  /// Switch for triggered / continuous readout
-  /// \param isContinuous - false for triggered readout, true for continuous readout
-  void setContinuousReadout(bool isContinuous);
+   /// Complete Clusterization
+   void FinishTask() override;
+
+   /// Switch for triggered / continuous readout
+   /// \param isContinuous - false for triggered readout, true for continuous readout
+   void setContinuousReadout(bool isContinuous);
 
   private:
-    bool mBoxClustererEnable;   ///< Switch to enable Box Clusterfinder
-    bool mHwClustererEnable;    ///< Switch to enable Hw Clusterfinder
-    bool mIsContinuousReadout;  ///< Switch for continuous readout
-    int mEventCount;            ///< Event counter
-    int mClusterSector = -1;    ///< Sector to be processed
-
-    std::unique_ptr<BoxClusterer> mBoxClusterer;    ///< Box Clusterfinder instance
-    std::unique_ptr<HwClusterer> mHwClusterer;      ///< Hw Clusterfinder instance
-
-    // Digit arrays
-    std::vector<o2::TPC::Digit> const* mDigitsArray[Sector::MAXSECTOR]; ///< Array of TPC digits
-    MCLabelContainer const*
-      mDigitMCTruthArray[Sector::MAXSECTOR]; ///< Array for MCTruth information associated to digits in mDigitsArrray
-
-    // Cluster arrays
-    std::vector<o2::TPC::Cluster> *mClustersArray;              ///< Array of clusters found by Box Clusterfinder
-    std::vector<o2::TPC::Cluster> *mHwClustersArray;            ///< Array of clusters found by Hw Clusterfinder
-    std::unique_ptr<MCLabelContainer> mClustersMCTruthArray;      ///< Array for MCTruth information associated to cluster in mClustersArrays
-    std::unique_ptr<MCLabelContainer> mHwClustersMCTruthArray;    ///< Array for MCTruth information associated to cluster in mHwClustersArrays
-
-    ClassDefOverride(ClustererTask, 1)
+   bool mIsContinuousReadout; ///< Switch for continuous readout
+   int mEventCount;           ///< Event counter
+   int mClusterSector = -1;   ///< Sector to be processed
+
+   std::unique_ptr<HwClusterer> mHwClusterer; ///< Hw Clusterfinder instance
+
+   // Digit arrays
+   std::shared_ptr<const std::vector<o2::TPC::Digit>> mDigitsArray; ///< Array of TPC digits
+   std::shared_ptr<const MCLabelContainer> mDigitMCTruthArray;      ///< Array for MCTruth information associated to digits in mDigitsArrray
```

Comment:
trying to understand: why a shared_ptr and not a unique_ptr?

Reply:
Because strictly speaking those pointers are shared with the HwClusterer and therefore it is not a unique_ptr. Maybe it would be even better, if the Process() function of the HwClusterer would take the digits and MCTruth as std::weak_ptr instead of std::shared_ptr?

Reply:
But aren't the digits owned by the `ClustererTask`?
Then one can simply make it a unique_ptr and pass the raw pointer from this to `mHwClusterer->Process` or do I miss something?

Reply:
yes, like this it is guaranteed that you get valid data.